### PR TITLE
[CEN-1539] Check for transaction integrity

### PR DESCRIPTION
**Pull request by TommasoLencioni**
#### Description

This PR adds an integrity check on transactions obtained from blobs.
Every field must match a regex pattern compliant with this documentation:
https://docs.pagopa.it/digital-transaction-register/v/digital-transaction-filter/acquirer-integration-with-pagopa-centrostella/integration/standard-pagopa-file-transactions

#### List of Changes

- Regex patterns annotation on Transaction fields.
- Malformed transactions test file added.
- Unit tests for the integrity check.

#### Motivation and Context

We can't take for granted the correctness of transaction fields because they can be generated by third party software.
This improvements let highlight malformed fields and avoids to pollute further steps of data manipulation.

#### How Has This Been Tested?

- [x] Unit Test Suite
- [ ] Integration Test Suite
- [ ] Api Tests
- [ ] Load Tests

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java [hunk 1]
```diff
@@ -22,58 +24,88 @@
...
+  @Pattern(regexp = "[0-9]{2}", message = "Circuit type length must match [0-9]{2}")
   String circuitType;
 
   @CsvBindByPosition(position = 3)
+  @Pattern(regexp = "[a-zA-Z0-9]{1,64}",
+      message = "HPAN length must be 64 alphanumeric alphanumeric char at max")
   String hpan;
 
+  //ISO8601
+  @NotNull
```
Comment (alberto-re-pagopa) [inline src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java:44]: Is there any downside in adding the @NotNull decorator on all non-nullable methods even if the @Pattern clearly prevent that? From a readability point-of-view could be an improvement, even if the implementation is already correct. What do you think?

Comment (TommasoLencioni) [reply src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java:44]: I deliberately removed every @NotNull annotation when a @Pattern annotation was present.
The ratio behind is that the validator, in presence of a null field, contained 2 violations (one for @NotNull and one for @Pattern) for the same field.
This behaviour polluted the log with a redundant reference to the same field.

I agree with you that specifying @NotNull is a clearer solution, but worsen the logging.
We could make a verbose violation evaluation step where we just collect a "distinct" of malformed fields, regardless of the source of the violation.

Comment (alberto-re-pagopa) [reply src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java:44]: In this case I think it's OK as it is, I was just wondering if there it was space for improving ;-)

File: src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java [hunk 2]
```diff
@@ -22,58 +24,88 @@
...
   @CsvBindByPosition(position = 1)
+  @Pattern(regexp = "[0-9]{2}", message = "Operation type length must match [0-9]{2}")
   String operationType;
 
   @CsvBindByPosition(position = 2)
+  @Pattern(regexp = "[0-9]{2}", message = "Circuit type length must match [0-9]{2}")
   String circuitType;
 
   @CsvBindByPosition(position = 3)
+  @Pattern(regexp = "[a-zA-Z0-9]{1,64}",
```
Comment (GiovanniMancini) [inline src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java:39]: Is the minimum number of character really 1 here?

File: src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java [hunk 3]
```diff
@@ -22,58 +24,88 @@
...
+  @Pattern(regexp = "[0-9]{6}|[0-9]{8}",
+      message = "Correlation ID length must match [0-9]{6}|[0-9]{8}")
   String bin;
 
   @CsvBindByPosition(position = 14)
+  @Pattern(regexp = "[a-zA-Z0-9]{1,5}", message = "MCC length must be 5 alphanumeric char at max")
   String mcc;
 
   @CsvBindByPosition(position = 15)
+  @Pattern(regexp = "[a-zA-Z0-9]{1,50}",
```
Comment (GiovanniMancini) [inline src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java:98]: Maybe this regex should be more restrictive

File: src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java [hunk 4]
```diff
@@ -22,58 +24,88 @@
...
+      message = "Fiscal code length must be 5 alphanumeric char at max")
   String fiscalCode;
 
   @CsvBindByPosition(position = 16)
+  @Pattern(regexp = "[a-zA-Z0-9]{0,50}",
```
Comment (GiovanniMancini) [inline src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java:103]: Maybe this regex should be more restrictive

**Review by GiovanniMancini — APPROVED** [review]

File: src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java (hunk 2 above)
Comment (TommasoLencioni) [reply src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java:39]: From the [documentation](https://docs.pagopa.it/digital-transaction-register/v/digital-transaction-filter/acquirer-integration-with-pagopa-centrostella/integration/standard-pagopa-file-transactions) the field is at max 64 alphanumeric char, but it is mandatory (so I avoided the 0 quantity).
I don't have enough information to put a lower limit of char for it.

File: src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java (hunk 3 above)
Comment (TommasoLencioni) [reply src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java:98]: Thank you, I'll fix it in another PR.

File: src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java (hunk 4 above)
Comment (TommasoLencioni) [reply src/main/java/it/gov/pagopa/rtd/ms/rtdmsingestor/model/Transaction.java:103]: Thank you, I'll fix it in another PR.

**Pull request merged**